### ISSUE TITLE
offset() output seems broken

### ISSUE THREAD

**Issue by ctrlcctrlv** [Logicalshift/flo_curves#1]
I'm surprised there are no issues.

Anyhow, I'm trying out `offset()`:

```rust
extern crate flo_curves;

use flo_curves::bezier;
use flo_curves::geo;

fn main() {
    let c = bezier::Curve{start_point: geo::Coord2(412.0,500.0), end_point: geo::Coord2(308.0,665.0), control_points: (geo::Coord2(163.0,589.0), geo::Coord2(163.0,504.0))};
    let o = bezier::offset(&c, 10.0, 10.0);

    println!("{:?}\n", &c);
    println!("{:?}\n", &o);

    print!(r#"<svg width="800" height="800"><path style="stroke:red;stroke-width:3px" d=""#);

    {
        let x = o[0].start_point.0;
        let y = o[0].start_point.1;
        print!("M {} {} ", x, y);
    }

    for cc in o {
        let x1 = cc.start_point.0;
        let y1 = cc.start_point.1;
        let x2 = cc.end_point.0;
        let y2 = cc.end_point.1;
        let h1x = cc.control_points.0.0;
        let h1y = cc.control_points.0.1;
        let h2x = cc.control_points.1.0;
        let h2y = cc.control_points.1.1;
        print!("C {} {}, {} {}, {} {} ", h1x, h1y, h2x, h2y, x2, y2);
    }

    println!(r#""></path></svg>"#);
}
```

Outputs:

![image](https://user-images.githubusercontent.com/838783/93004679-0e7a7b80-f4fe-11ea-9715-0033dfd3bb6c.png)

Which, suffice to say, isn't at all what I expected. What am I not understanding? :)

**Comment by Logicalshift** [comment]
Always has to be a first issue :-)

I've been meaning to revamp `offset()` for a while as it doesn't behave well when a curve starts to form a cusp - as this example does, and I have to confess I've seen this bug before but attributed it to the problems around cusps, but now I've looked at the issue a bit more and this is a bug around splitting up a certain class of curve with double inflection points: that should be fixed in the latest version in the repository now (single character typo - see ffa5fa4...)

With this fix, this curve now looks like this (I've added the original for comparison):

<img width="253" alt="Screenshot 2020-09-12 at 23 01 57" src="https://user-images.githubusercontent.com/4084632/93005766-0fd29500-f54c-11ea-80a1-bfacb205c992.png">

The slight 'lumpiness' is due to the problem with cusps I mentioned.

**Comment by ctrlcctrlv** [comment]
Nice! :-)

I don't want to presume anything, but if you're interested in being commissioned, please do email me (email in bio). The open source font editor project I'm working on, [MFEQ](https://github.com/mfeq/Qglif), has some financial backing and you'd be perfect I think, and this library is just what I need I do think.

**Comment by ctrlcctrlv** [comment]
![image](https://user-images.githubusercontent.com/838783/93007054-d6ccfd00-f518-11ea-9276-c5accd5bbcd4.png)

The breakage is not nearly so bad anymore, but even on what I consider a very easy case it still exists, so I'll leave this open if you don't mind :-)

```rust
extern crate flo_curves;

use flo_curves::bezier;
use flo_curves::geo;

fn svgify(o: &Vec<bezier::Curve<geo::Coord2>>, stroke_color: &str) {
    print!(r#"<path style="fill:none;stroke:{};stroke-width:3px" d=""#, stroke_color);

    {
        let x = o[0].start_point.0;
        let y = o[0].start_point.1;
        print!("M {} {} ", x, y);
    }

    for cc in o {
        let x1 = cc.start_point.0;
        let y1 = cc.start_point.1;
        let x2 = cc.end_point.0;
        let y2 = cc.end_point.1;
        let h1x = cc.control_points.0.0;
        let h1y = cc.control_points.0.1;
        let h2x = cc.control_points.1.0;
        let h2y = cc.control_points.1.1;
        print!("C {} {}, {} {}, {} {} ", h1x, h1y, h2x, h2y, x2, y2);
    }

    println!(r#""></path>"#);
}

fn main() {
    let c = bezier::Curve{start_point: geo::Coord2(110.0,110.0), end_point: geo::Coord2(500.0,110.0), control_points: (geo::Coord2(110.0,300.0), geo::Coord2(500.0,300.0))};
    let o = bezier::offset(&c, 4.0, 40.0);
    let o2 = bezier::offset(&c, -4.0, -40.0);

    println!("{:?}\n", &c);
    println!("{:?}\n", &o);

    println!(r#"<svg width="800" height="800">"#);
    svgify(&o, "red");
    svgify(&o2, "purple");
    svgify(&vec![c], "blue");
    println!(r#"</svg>"#);
}
```

Happy to commission a fix if that's something you're interested in.

**Comment by Logicalshift** [comment]
This new problem turns out to be that I was calling `line_intersects_line()` where I should have been calling `ray_intersects_ray` so it was using the wrong algorithm most of the time. I've corrected that and some other bugs I found in the latest revision. Here's how your example looks with the latest version.

![path10](https://user-images.githubusercontent.com/4084632/93944567-fb04b700-fd2c-11ea-952e-cfbdce14c598.png)

There are still some curves that produce issues, though they're less noticeable than they were before. Here's one I found as an example (note that it's a very small range of values that produce issues):

```
let c = Curve::from_points(Coord2(516.170654296875, 893.27001953125), (Coord2(445.1522921545783, 856.2028149461783), Coord2(447.7831664737134, 878.3276285260063)), Coord2(450.51018453430754, 901.260980294519));
```

As to commissions: presently I'm working full time, so I'm afraid I'm not able to take any on at the moment.
